### DISCUSSION TITLE
Module support for better functional programming

### DISCUSSION THREAD

**Discussion by ziaulhasanhamim** [dotnet/csharplang#5034] [Language Ideas]
In the last few years, more and more functional features are added to c#. you cant directly create functions in c#. You have to create a static class.
```c#
using System;

namespace Application
{
    public static class Helper
    {
       private static int counter = 0;
       public static void IncrementAndPrint(int count)
       {
           counter += count;
           Console.WriteLine($"counter {count}");
       }
    }
}
```
With modules, this piece of code can be simplified a lot. 
```c#
mod Application.Helper;

using System;

var counter = 0; // you can declare private static field with var;
void IncrementAndPrint(int count) // methods are by default public and static in modules
{
    counter += count;
    Console.WriteLine($"counter {count}");
}
```
Modules will also be useful in the new asp.net core minimal APIs

**Comment by svick** [comment]
You seem to be proposing several things:

1. The ability to declare a type at the file level, which is a duplicate of https://github.com/dotnet/csharplang/discussions/928.
2. The ability to have `var` fields, which has been previously discussed at https://github.com/dotnet/csharplang/discussions/2828. This is a problematic feature and is unlikely to ever happen.
3. Different (and inconsistent) default modifiers for members of modules. I don't see the big advantage of this. It might make your code a bit shorter, but also harder to understand. And the inconsistency (some members seem to be private by default, others public) is not appealing either.

Overall, I don't really see how this would lead to "better functional programming".

**Reply by ziaulhasanhamim** [reply]
the thing is I was proposing the default access modifier changes and var fields for only modules. this is very similar to f# modules. modules will help to declare functions and variables without class or structs. thus brings more functional prgramming
